### DISCUSSION TITLE
React - Redux - Practice Exercices

### DISCUSSION THREAD

**Discussion by akash-coded** [akash-coded/mern#72] [Tasks]
1. Create a plain React app displaying a “Room” with a lightswitch. Clicking the button toggles the state, which turns the light on and off.

    Everything can be in one file: _index.js_

    Your task is to add Redux to this app and move the lightswitch state out of the React component and into Redux.

    Once you’re done, clicking the button should dispatch an action, and the reducer should return the updated lightswitch state.

    You can leave everything in one file, or split out the Room component into its own file if you like.

    If you do leave it all in index.js, you won’t need to export the connected Room component. Instead, save it in a separate variable, like `const ConnectedRoom = connect(mapStateToProps)(Room)`

2. Build a Step Counter. It’s just like a Fitbit, except entirely manual.
    
    ![image](https://user-images.githubusercontent.com/60097560/179269235-b17a4a4e-80e9-4bce-ae2f-31641186d248.png)

    Create a brand new React project with [Create React App 2](https://github.com/facebook/create-react-app) or [CodeSandbox 4](https://codesandbox.io/s/new). Delete the boilerplate code from index.js and start fresh.

    Add Redux to the project as you’ve done before.

    There’s just one component here: the StepCounter. It should have 2 buttons, one “Add Step” and one “Reset Steps”, as shown above. Clicking “Add Step” should increment the step counter by 1, and “Reset Steps” should clear it. (think: how many actions will you need?)

    For each action, create an action constant and an action creator. Use mapDispatchToProps to pass the actions in as props to the StepCounter component. Try out the object form of mapDispatch as well as the function form.

    You can break the project up into files if you like, or keep it all in one file. If you want a suggestion, try for an index.js with the createStore + reducer, a StepCounter.js with the component and any necessary connect() setup, and an actions.js with the actions (don’t forget to export them!).

    If you want to make it pretty, go for it. But it’s fine to leave it unstyled like this: 
    ![image](https://user-images.githubusercontent.com/60097560/179269483-147f2831-c388-48e9-9896-b9f788adda4d.png)

**Comment by skagarwal5555** [comment]
[https://github.com/skragarwaltcs/mern/tree/master/front-end/redux/todolist](url)

![image](https://user-images.githubusercontent.com/105811212/179277489-e9007211-d292-44af-a025-3da2cd660d4f.png)

**Reply by skagarwal5555** [reply]
[https://github.com/skragarwaltcs/mern/tree/master/front-end/redux/lightswitch](url)

![image](https://user-images.githubusercontent.com/105811212/179289263-96f81a27-0f7b-471c-a79d-058746899ef5.png)

**Comment by Amrita-12345** [comment]
## Q.1. with react
## App.js
```js
import './App.css';
import {useState} from 'react'
import React from 'react';
function App() {
const [light,setLight]=useState('off')
  const toggleChange=()=>
  {
   if(light==='off')
   {
 setLight('on')
 }
 else
 {
   setLight('off')
 }
  }
return (
    <div className="App">
 <h1>
      Room with lightswitch:{`${light}`}
   </h1>
        <button  onClick={toggleChange}>Click </button>
 </div>
  );
}
export default App;
```
### screenshot
![image](https://user-images.githubusercontent.com/104146616/189912701-e4c48f3d-5090-4e57-b1e4-83b39ae70cd4.png)
![image](https://user-images.githubusercontent.com/104146616/189912775-fca74750-37a7-47ed-9ae5-64d96049970e.png)


## Q.2 with redux
## index.js
```js
import React from 'react';
import ReactDOM from 'react-dom/client';
import './index.css';
import App from './App';
import reportWebVitals from './reportWebVitals';
import { store } from './store';
import { Provider } from 'react-redux';
 
store.subscribe(()=>console.log(store.getState()));
const root = ReactDOM.createRoot(document.getElementById('root'));
root.render(
  
  <React.StrictMode>
    <Provider store={store}>
   <App/>
   </Provider>
  </React.StrictMode>
);
reportWebVitals();
```
## App.js
```js
import './App.css';
import {useState} from 'react'
import React from 'react';
import {useSelector,useDispatch} from  'react-redux'
import { addStep,resetStep } from './actions/index';
function App() {
const myState=useSelector((state)=>state.addStep);
        const dispatch=useDispatch();
  return (
    <div className="App">
       <h1>You've walked {myState} steps today! </h1>
        <button type="button" onClick={()=>dispatch(addStep())} className="btn btn-primary btn-lg my-2">Add a step</button>
        <br></br>
        <button type="Reset" onClick={()=>dispatch(resetStep())} className="btn btn-secondary btn-lg my-2">Reset steps</button>
        </div>
  );
}
export default App;
```
## add.js
```js
export const addStep=(state=0,action)=>
{
switch(action.type)
{
    case  "ADDSTEP": return state+1;
    case "RESETSTEP": return 0;
    default:return state;
}
}
```
## reducer-index.js
```js
import { addStep } from "./add";
import { combineReducers } from "redux";
const rootReducer = combineReducers( 
    {addStep});
    export default rootReducer;
```
## action-index.js
```js
export const addStep=()=>
    {
        return{
type: "ADDSTEP",
    }
    }
    export const resetStep=()=>
    {
        return{
type: "RESETSTEP",
    }
    }
```
## store.js
```js
import { createStore } from "redux";
import rootReducer from "./reducers/index";
 export const store = createStore(rootReducer);
```

### Screenshot
![image](https://user-images.githubusercontent.com/104146616/189908275-18b2497e-ec52-4abe-8197-a960ee830961.png)

**Comment by shubhamk-coder** [comment]
## Question-1 `with React`
### App.js
```javascript
import "./App.css";
import { useState } from "react";
import React from "react";

const roomLight = document.getElementById("room");
function App() {
  const [light, setLight] = useState("OFF");
  const lightChange = () => {
    if (light === "OFF") {
      roomLight.classList.add("active");
      setLight("ON");
    } else {
      roomLight.classList.remove("active");
      setLight("OFF");
    }
  };
  return (
    <div className="App">
      <h1 id="room">Room with a Light Switch : {`${light}`}</h1>
      <button onClick={lightChange}>Click</button>
    </div>
  );
}
export default App;
```
### App.css
```css
button {
  height: 20px;
  width: 60px;
  border-radius: 10%;
  border-color: aqua;
}

#room.active {
  color: aqua
}

#room {
  color: black;
}
```
### Screenshots
![image](https://user-images.githubusercontent.com/36131539/190244216-37e57f85-5e18-4e5c-806d-e5e975167214.png)
![image](https://user-images.githubusercontent.com/36131539/190244258-152c470c-a23e-47d9-a58c-e392f4c23db0.png)
## Question - 2 `with Redux`
### index.js
```javascript
import React from "react";
import ReactDOM from "react-dom/client";
import "./index.css";
import App from "./App";
import reportWebVitals from "./reportWebVitals";
import { store } from "./Store";
import { Provider } from "react-redux";

store.subscribe(() => console.log(store.getState()));
const root = ReactDOM.createRoot(document.getElementById("root"));
root.render(
  <React.StrictMode>
    <Provider store={store}>
      <App />
    </Provider>
  </React.StrictMode>
);
reportWebVitals();
```
### App.js
```javascript
import "./App.css";
// eslint-disable-next-line no-unused-vars
import { useState } from "react";
import React from "react";
import { useSelector, useDispatch } from "react-redux";
import { addStep, resetStep } from "./Action";
function App() {
  const myState = useSelector((state) => state.addStep);
  const dispatch = useDispatch();
  return (
    <div className="App">
      <h1>
        You've walked <span className="span">{myState}</span> steps today!{" "}
      </h1>
      <button
        className="button"
        type="button"
        onClick={() => dispatch(addStep())}
      >
        Add step
      </button>
      <br></br>
      <button
        className="button"
        type="Reset"
        onClick={() => dispatch(resetStep())}
      >
        Reset steps
      </button>
    </div>
  );
}
export default App;
```
### Action.js
```javascript
export const addStep = () => {
  return {
    type: "AddStep",
  };
};
export const resetStep = () => {
  return {
    type: "ResetStep",
  };
};
```
### Add.js
```javascript
export const addStep = (state = 0, action) => {
  switch (action.type) {
    case "AddStep":
      return state + 1;
    case "ResetStep":
      return 0;
    default:
      return state;
  }
};
export default addStep;
```
### Reducer.js
```javascript
import { addStep } from "./Add";
import { combineReducers } from "redux";
const rootReducer = combineReducers({ addStep });
export default rootReducer;
```
### Store.js
```javascript
import { createStore } from "redux";
import rootReducer from "./Reducer";
export const store = createStore(rootReducer);
```
### App.css
```css
.App {
  text-align: center;
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: center;
}
.button {
  margin-top: 10px;
  margin-left: 10px;
  height: 30px;
  border-radius: 5px;
  border-color: aqua;
}
.span {
  color: aqua;
}
```
### Screenshots
![image](https://user-images.githubusercontent.com/36131539/190256001-2d9fc1c2-0531-49d7-9fa4-334e7111418f.png)
![image](https://user-images.githubusercontent.com/36131539/190256063-5b709801-1d1e-4492-ba3a-f304a9b38567.png)
![image](https://user-images.githubusercontent.com/36131539/190256198-73a169c1-c70a-47c6-a1e1-a9d9a2eb63b0.png)

**Comment by Neethumar** [comment]
### Question 1 with React

```javascript
import React, { useState } from 'react'

function App() {
    const [switchState, setSwitchState] = useState(false);

    return (
        <>
            {switchState ? <img src="https://thumbs.dreamstime.com/b/vector-light-bulb-white-file-eps-format-34983265.jpg" width="90" alt="img1"/> : <img src="https://previews.123rf.com/images/ericmilos/ericmilos0912/ericmilos091200136/6109526-3d-render-of-light-bulb-on-white.jpg" width="90" alt="img2"/>}
            <br />
            <button onClick={() => setSwitchState(!switchState)}>Switch {switchState ? "OFF" : "ON"}</button>
        </>
    )
}

export default App

```
<img width="187" alt="image" src="https://user-images.githubusercontent.com/93718958/190360414-c6b5280b-4f6c-46a1-97e2-6013be3f383f.png">

<img width="113" alt="image" src="https://user-images.githubusercontent.com/93718958/190360508-a2794a47-4ad4-4032-aaaf-865784ccae19.png">

### Question 2 with Redux

### index.js
```javascript
import React from 'react';
import ReactDOM from 'react-dom/client';
import './index.css';
import App from './App';
import reportWebVitals from './reportWebVitals';
import { store } from './Store';
import { Provider } from 'react-redux';
 
store.subscribe(()=>console.log(store.getState()));
const root = ReactDOM.createRoot(document.getElementById('root'));
root.render(
  
  <React.StrictMode>
    <Provider store={store}>
   <App/>
   </Provider>
  </React.StrictMode>
);
reportWebVitals();
```
### App.js

```javascript
import './App.css';
import React from 'react';
import { useSelector, useDispatch } from 'react-redux'
import { addStep, resetStep } from './Action';
function App() {
  const myState = useSelector((state) => state.addStep);
  const dispatch = useDispatch();
  return (
    <div className="App">
      <h1>You've walked {myState} steps today! </h1>
      <button type="button" onClick={() => dispatch(addStep())}>Add a step</button>
      <br></br>
      <button type="Reset" onClick={() => dispatch(resetStep())} >Reset steps</button>
    </div>
  );
}
export default App;
```

### Add.js
```javascript
export const addStep = (state = 0, action) => {
    switch (action.type) {
        case "ADDSTEP":
            return state + 1;
        case "RESETSTEP":
            return 0;
        default:
            return state;
    }
}
```

### Action.js

```javascript
export const addStep = () => {
    return {
        type: "ADDSTEP",
    }
}
export const resetStep = () => {
    return {
        type: "RESETSTEP",
    }
}
```

### Reducer.js
```javascript

import { addStep } from "./Add";
import { combineReducers } from "redux";

const rootReducer = combineReducers({ addStep });

export default rootReducer;
```
<img width="426" alt="image" src="https://user-images.githubusercontent.com/93718958/190363117-b6df6fc6-538c-47db-ad1d-4a7af50b4f67.png">
<img width="356" alt="image" src="https://user-images.githubusercontent.com/93718958/190363212-c1528031-7d3c-48e3-9bcb-e1ba7bfd1209.png">

**Comment by Avinash-chengot** [comment]
app.js

import "./styles.css";

export default function App() {

  const status="On";
  return (
    <div className="App">
      <h1>Light {status}</h1>
      <button Onclick="stch()">Switch</button>
          </div>
  );

  
 
}

function stch(){
  document.querySelector(".App").style.color="red"
}
``
